### ISSUE TITLE
PEG implementation for R

### ISSUE THREAD

**Issue by sckott** [ropensci/unconf#14]
There are [Parsing expression grammars (PEGs)](https://en.wikipedia.org/wiki/Parsing_expression_grammar) implementations in various languages, including [Ruby's treetop](https://github.com/nathansobo/treetop)

A PEG implementation in R could be super useful for me at least. The use case I have is that right now in [taxize](https://github.com/ropensci/taxize) we use a web API to parse scientific names, which are often very hard to parse. The web API is Ruby based and uses the `treetop` gem (the API [GlobalNamesArchitecture/biodiversity](https://github.com/GlobalNamesArchitecture/biodiversity)). It does a very nice job, but it would be nice to have a native R PEG so that we don't have to use a web API. 

I imagine there are other use cases. 

If this is already done in R, awesome, where's it at?

**Comment by richfitz** [comment]
If it's not done, then porting [boost spirit](http://www.boost.org/doc/libs/1_57_0/libs/spirit/doc/html/spirit/abstracts/parsing_expression_grammar.html) with Rcpp is one option that could be fun.

**Comment by sckott** [comment]
Thanks Rich. I'll take a look

**Comment by hadley** [comment]
There's going to be a pretty strong mismatch between compile time boost::spirit and run-time R. I suspect it might be easier to solve with boost::spirit for specific grammars.

**Comment by sckott** [comment]
thanks @hadley - make sense

**Comment by richfitz** [comment]
I guess then the decision would be: build off the boost one and provide nice interfaces for people to add their own grammars (in the approach of some of the header-only Rcpp libraries - mostly focussing on things like nice utility functions for doing type coercion) or port a PEG from a dynamic language.
